File: movf/bvh/BVH.m
classdef BVH < handle
  properties
    all_BVHNodes BVHNode
    nodes_used (1,1) uint64 = 0
    mesh Mesh
    mesh_element_indices (1,:) uint64
    minimum_triangle_threshold (1,1) uint64 = 2
    num_cost_evaluation_points (1,1) uint64 = 4
  end

  methods (Access = public)
    function obj = BVH(~)
    end

    function linkToMesh(obj, mesh)
      arguments
        obj BVH
        mesh Mesh
      end
      obj.mesh = mesh;
      obj.mesh_element_indices = linspace(1, mesh.numElements, mesh.numElements);
      obj.all_BVHNodes(obj.maxNumNodes) = BVHNode;
    end

    function setMinimumNumTriangles(obj, min_triangles)
      arguments
        obj BVH
        min_triangles (1,1) uint64
      end
      obj.minimum_triangle_threshold = min_triangles;
    end

    function setNumCostEvaluationPoints(obj, num_cost_evaluation_points)
      arguments
        obj BVH
        num_cost_evaluation_points (1,1) uint64
      end
      obj.num_cost_evaluation_points = num_cost_evaluation_points;
    end

    function constructBVH(obj)
      fprintf('Constructing BVH for mesh with %d elements\n', obj.mesh.numElements)
      root_node_index = 1;
      obj.nodes_used = 1;
      root_node = obj.all_BVHNodes(root_node_index);
      root_node.growToIncludeMesh(obj.mesh);
      root_node.setNumTriangles(obj.mesh.numElements);
      obj.subdivideNode(root_node_index);
    end

    function intersectRayWithBVH(obj, ray)
      arguments
        obj BVH
        ray Ray
      end
      obj.intersectRayWithBVHNode(ray, 1);
    end
    
    function plotMesh(obj)
      obj.mesh.plotMesh
      obj.plotNodeBBox(1);
    end

    function plotLeafNodes(obj)
      for i = 1 : obj.maxNumNodes
        if obj.all_BVHNodes(i).getNumTriangles > 0
          obj.plotNodeBBox(i);
          obj.plotNodeSubMesh(i);
        end
      end
    end
  end

  methods (Access = private)
    function max_num_nodes = maxNumNodes(obj)
      max_num_nodes = obj.mesh.numElements * 2 - 1;
    end

    function subdivideNode(obj, node_index)
      arguments
        obj BVH
        node_index (1,1) uint64
      end
      fprintf('Subdividing Node %d\n', node_index)

      current_node = obj.all_BVHNodes(node_index);
      %! skip subdivision if the current iteration contains too few triangles
      if current_node.getNumTriangles <= obj.minimum_triangle_threshold
        fprintf('Not splitting node %d, number of triangles lower than threshold\n', node_index)
        return
      end

      %! [1] determine the cardinal axis and position along it to split the bounding box
      [axis_index, location_on_axis_to_split, best_cost] = obj.getSplitLocation(node_index);
      %! skip subdivision if the current iteration is already lower cost than a possible subdivision
      if current_node.getNodeCost <= best_cost
        fprintf('RETURN || splitting node %d provides no decrease in cost\n', node_index)
        return
      end

      %! [2] split the group of primitives in two halves using the split plane
      split_index = obj.splitPrimitives(node_index, axis_index, location_on_axis_to_split);

      %! skip subdivision if the "split" contains either no triangles or all the triangles in the current node
      num_triangles_on_left = split_index - current_node.getFirstTriangleIndex;
      if num_triangles_on_left == 0 || num_triangles_on_left == current_node.getNumTriangles
        fprintf('RETURN || somehow one side has all the triangles\n')
        return
      end

      %! [3] create child nodes for each half
      left_child_index = obj.createChildNodes(node_index, split_index, num_triangles_on_left);

      %! [4] recurse into each of the child nodes
      obj.subdivideNode(left_child_index);
      obj.subdivideNode(left_child_index + 1);
    end

    function [axis_index, best_position, best_cost] = getSplitLocation(obj, node_index)
      arguments
        obj BVH
        node_index (1,1) uint64
      end
      fprintf('Evaluating optimal split location\n')
      current_node = obj.all_BVHNodes(node_index);
      
      node_span = current_node.getAABBSpan;
      axis_index = node_span.getLongestDirection;
      span_on_longest_direction = node_span.getComponentByAxis(axis_index);
      
      minimum_value_on_axis = current_node.getAABBMin.getComponentByAxis(axis_index);

      num_evaluation_points = obj.num_cost_evaluation_points;
      evaluation_point_costs = zeros(1, num_evaluation_points);
      evaluation_point_positions = zeros(1, num_evaluation_points);
      node_submesh = obj.extractNodeSubMesh( node_index );
      for i = 1 : num_evaluation_points
        evaluation_point_positions(i) = minimum_value_on_axis + span_on_longest_direction * (double(i) / (double(num_evaluation_points) + 1));
        evaluation_point_costs(i) = evaluateNodeChildrenSurfaceAreaHeuristic(current_node, node_index, node_submesh, axis_index, evaluation_point_positions(i));
      end

      [best_cost, best_cost_index] = min(evaluation_point_costs);
      best_position = evaluation_point_positions(best_cost_index);
    end

    function index_at_which_to_split = splitPrimitives(obj, node_index, axis_index, split_location)
      arguments
        obj BVH
        node_index (1,1) uint64
        axis_index (1,1) uint8
        split_location (1,1) double
      end
      fprintf('Splitting primitives on axis %d at %.2f\n', axis_index, split_location)
      current_node = obj.all_BVHNodes(node_index);
      index_at_which_to_split = current_node.getFirstTriangleIndex;
      index_of_last_unsorted_element = index_at_which_to_split + current_node.getNumTriangles - 1;
      while index_at_which_to_split <= index_of_last_unsorted_element
        mesh_element = obj.mesh.getElement(obj.mesh_element_indices(index_at_which_to_split));
        if mesh_element.getCentroid.getComponentByAxis(axis_index) < split_location
          index_at_which_to_split = index_at_which_to_split + 1;
        else
          obj.swapElements(index_at_which_to_split, index_of_last_unsorted_element);
          index_of_last_unsorted_element = index_of_last_unsorted_element - 1;
        end
      end
    end

    function swapElements(obj, index_one, index_two)
      arguments
        obj BVH
        index_one (1,1) uint64
        index_two (1,1) uint64
      end
      value_at_index_one = obj.mesh_element_indices(index_one);
      value_at_index_two = obj.mesh_element_indices(index_two);
      obj.mesh_element_indices(index_one) = value_at_index_two;
      obj.mesh_element_indices(index_two) = value_at_index_one;
    end

    function left_child_index = createChildNodes(obj, node_index, split_index, num_triangles_on_left)
      arguments
        obj BVH
        node_index (1,1) uint64
        split_index (1,1) uint64
        num_triangles_on_left (1,1) uint64
      end
      fprintf('Preparing the children nodes of node %d\n', node_index)
      fprintf('<> prior to creation of children nodes, num_triangles_on_left = %d <>\n', num_triangles_on_left)
      current_node = obj.all_BVHNodes(node_index);

      left_child_index = obj.nodes_used + 1;
      obj.nodes_used = left_child_index + 1;

      current_node.setChildOneIndex(left_child_index);
      
      obj.all_BVHNodes(left_child_index).setFirstTriangleIndex(current_node.getFirstTriangleIndex);
      obj.all_BVHNodes(left_child_index).setNumTriangles(num_triangles_on_left);
      
      obj.all_BVHNodes(left_child_index + 1).setFirstTriangleIndex(split_index);
      obj.all_BVHNodes(left_child_index + 1).setNumTriangles(current_node.getNumTriangles - num_triangles_on_left);
      
      fprintf('<> after creation of children nodes, %d elements on left, %d elements on right\n', ...
        num_triangles_on_left, ...
        current_node.getNumTriangles - num_triangles_on_left)

      current_node.setNumTriangles(0);

      obj.constructNewNode(left_child_index);
      obj.constructNewNode(left_child_index + 1);
    end

    function constructNewNode(obj, node_index)
      arguments
        obj BVH
        node_index (1,1) uint64
      end
      fprintf('Constructing new node %d\n', node_index)
      current_node = obj.all_BVHNodes(node_index);
      fprintf('Node being constructed should contain %d triangles\n', current_node.getNumTriangles)
      fprintf('Submesh being extracted has %d elements\n', length( obj.getSubMeshIndices(node_index) ))
      current_node.growToIncludeMesh( obj.mesh.getSubMesh( obj.getSubMeshIndices(node_index) ) );
    end

    function intersectRayWithBVHNode(obj, ray, node_index)
      arguments
        obj BVH
        ray Ray
        node_index (1,1) uint64
      end
      fprintf('Intersection Test\n')
      current_node = obj.all_BVHNodes(node_index);
      if ~current_node.intersectRayWithNodeBBox(ray)
        return;
      end
      if current_node.isLeaf
        for i = current_node.getFirstTriangleIndex : current_node.getFirstTriangleIndex + current_node.getNumTriangles - 1
          ray.triangleIntersection(obj.mesh.getElement(obj.mesh_element_indices(i)));
        end
      else
        obj.intersectRayWithBVHNode(ray, current_node.child_one_index);
        obj.intersectRayWithBVHNode(ray, current_node.child_one_index + 1);
      end
    end    

    function plotNodeSubMesh(obj, node_index)
      arguments
        obj BVH
        node_index (1,1) uint64
      end
      m = obj.extractNodeSubMesh(node_index);
      m.plotMesh;
    end

    function sub_mesh = extractNodeSubMesh(obj, node_index)
      arguments
        obj BVH
        node_index (1,1) uint64
      end
      sub_mesh = obj.mesh.getSubMesh( obj.getSubMeshIndices(node_index) );
    end

    function plotNodeBBox(obj, node_index)
      arguments
        obj BVH
        node_index (1,1) uint64
      end
      b = obj.all_BVHNodes(node_index);
      AABBmin = b.getAABBMin;
      AABBmax = b.getAABBMax;
      all_min = [AABBmin.getX, AABBmin.getY, AABBmin.getZ];   %! all coords are minimum
      yz_x = [AABBmax.getX, AABBmin.getY, AABBmin.getZ];      %! yz min, x max
      z_xy = [AABBmax.getX, AABBmax.getY, AABBmin.getZ];      %! z min, xy max
      xz_y = [AABBmin.getX, AABBmax.getY, AABBmin.getZ];      %! xz min, y max
      xy_z = [AABBmin.getX, AABBmin.getY, AABBmax.getZ];      %! xy min, z max
      y_xz = [AABBmax.getX, AABBmin.getY, AABBmax.getZ];      %! y min, xz max
      all_max = [AABBmax.getX, AABBmax.getY, AABBmax.getZ];   %! all coords are maximum
      x_yz = [AABBmin.getX, AABBmax.getY, AABBmax.getZ];      %! x min, yz max
      vertices = [all_min; yz_x; z_xy; xz_y; xy_z; y_xz; all_max; x_yz];
      faces = [1,2,3,4; ... %! construct all faces by indexing through vertices
               1,2,6,5; ...
               1,4,8,5; ...
               2,3,7,6; ...
               3,4,8,7; ...
               5,6,7,8];
      if b.getNumTriangles > 0
        face_color = 'red';
        face_alpha = 0.3;
      else
        face_color = 'green';
        face_alpha = 0.1;
      end
      patch('Faces', faces, 'Vertices', vertices, 'FaceColor', face_color, 'FaceAlpha', face_alpha);
    end

    function submesh_indices = getSubMeshIndices(obj, node_index)
      arguments
        obj BVH
        node_index (1,1) uint64
      end
      current_node = obj.all_BVHNodes(node_index);
      submesh_indices = obj.mesh_element_indices(current_node.getElementArraySubindices);
    end
  end

  methods (Static)
  end
end

File: ovf/geometry/src/BVH.cpp
#include "BVH.hpp"

namespace openviewfactor {

  //* ------------------------------ PUBLIC METHODS ------------------------------ *//

  template <typename t>
  gpuify BVH<t>::BVH()
    : _nodes(std::vector<BVHNode<t>>()), _triangulations(std::vector<size_t>()), _points(std::vector<t>()), _mesh_element_indices(std::vector<size_t>()), _nodes_used(0), _minimum_triangle_threshold(2), _num_cost_evaluation_points(4) {}

  template <typename t>
  gpuify BVH<t>& BVH<t>::linkToTriangulation(std::vector<size_t> &triangulations, std::vector<t> &points) {
    if (this->isLinked()) {
      _triangulations.clear();
      _points.clear()
      _mesh_element_indices.clear();
      _nodes.clear();
      _nodes_used = 0;
    }
    _triangulations = triangulations;
    _points = points;
    _mesh_element_indices = std::vector<size_t>(triangulations.size()/3);
    std::iota(_mesh_element_indices.begin(), _mesh_element_indices.end(), 0);
    _nodes = std::vector<BVHNode<t>>(this->getMaxNumNodes());
    return *this;
  }

  template <typename t>
  gpuify bool BVH<t>::isLinked() const { return !(*_triangulation == Triangulation<t>()); }

  template <typename t>
  gpuify BVH<t>& BVH<t>::setMinimumNumTriangles(unsigned int min_triangles) {
    _minimum_triangle_threshold = min_triangles;
    return *this;
  }
  template <typename t>
  gpuify BVH<t>& BVH<t>::setNumCostEvaluationPoints(unsigned int num_cost_evaluation_points) {
    _num_cost_evaluation_points = num_cost_evaluation_points;
    return *this;
  }

  template <typename t>
  gpuify BVH<t>& BVH<t>::constructBVH() {
    if (!(this->isLinked())) {
      throw std::runtime_error("BVH is not linked to a triangulation. Cannot construct the BVH");
    }
    std::cout << "> BVH Construction Initiated" << std::endl;
    unsigned int root_node_index = 0;
    _nodes[root_node_index].growToIncludeTriangulation(_triangulation);
    this->setNumNodesUsed(1);
    this->subdivideNode(root_node_index);
    std::cout << "> BVH Construction Completed. " << this->getNumNodesUsed() << " Nodes used." << std::endl;
    return *this;
  }

  template <typename t>
  gpuify void BVH<t>::intersectRayWithBVH(std::shared_ptr<Ray<t>> ray) {
    this->intersectRayWithBVHNode(ray, 0);
  }

  template <typename t>
  gpuify unsigned int BVH<t>::getNumNodesUsed() { return _nodes_used; }

  template <typename t>
  gpuify void BVH<t>::writeToFile(const std::string& filename) const {
    std::ofstream outfile(filename + ".txt");
    outfile << "Encoding,X,Y,Z\n";
    unsigned int j = 0;
    for (unsigned int i = 0; i < _nodes_used; i++) {
      auto node = _nodes[i];
      if (node.isLeaf()) {
        node.writeToFile(outfile, j++);
      }
    }
    outfile.close();
  }

  //* ------------------------------ PRIVATE METHODS ------------------------------ *//

  template <typename t>
  gpuify BVH<t>& BVH<t>::setNumNodesUsed(unsigned int n) {
    _nodes_used = n;
    return *this;
  }

  template <typename t>
  gpuify unsigned int BVH<t>::getMaxNumNodes() const {
    if (!(this->isLinked())) {
      throw std::runtime_error("BVH is not linked to a triangulation. Cannot evaluate the maximum number of elements");
    }
    return (2 * (_triangulation->getNumElements()) - 1);
  }

  template <typename t>
  gpuify BVH<t>& BVH<t>::subdivideNode(unsigned int node_index) {
    //* --------------- FUNCTION ACTIONS --------------- *//
    //* [1] skip subdivision if current iteration contains too few triangles
    //* [2] determine cardinal axis, position along it, and cost for splitting the node
    //* [3] skip subdivision if current iteration is already lower cost than a possible subdivision
    //* [4] split the group of primitives in two halves using the split plane
    //* [5] skip subdivision if the "split" contains either no triangles or all the triangles
    //* [6] create child nodes for each half of the current node
    //* [7] recurse into each of the child nodes

    //* --------------- [1] --------------- *//
    if ((_nodes[node_index]).getNumTriangles() <= _minimum_triangle_threshold) { return *this; }

    //* --------------- [2] --------------- *//
    unsigned int axis_index = (_nodes[node_index]).getSplitLocationAxis();
    std::pair<t, t> location_and_cost = (_nodes[node_index]).getBestSplitLocationAndCost(_triangulation, _mesh_element_indices, axis_index, _num_cost_evaluation_points);
    t best_location = location_and_cost.first;
    t best_cost = location_and_cost.second;
    //* --------------- [3] --------------- *//
    if ((_nodes[node_index]).getNodeCost() <= best_cost) { return *this; }

    //* --------------- [4] --------------- *//
    unsigned int split_index = this->splitPrimitives(node_index, axis_index, best_location);
    //* --------------- [5] --------------- *//
    unsigned int num_triangles_on_left = split_index - (_nodes[node_index]).getFirstTriangleIndex();
    if (num_triangles_on_left == 0 || num_triangles_on_left == (_nodes[node_index]).getNumTriangles()) { return *this; }

    //* --------------- [6] --------------- *//
    unsigned int left_child_index = this->createChildNodes(node_index, split_index, num_triangles_on_left);
    //* --------------- [7] --------------- *//
    this->subdivideNode(left_child_index);
    this->subdivideNode(left_child_index + 1);

    return *this;
  }

  template <typename t>
  gpuify unsigned int BVH<t>::splitPrimitives(unsigned int node_index, unsigned int axis_index, t split_location) {
    
    unsigned int index_at_which_to_split = (_nodes[node_index]).getFirstTriangleIndex();
    unsigned int index_of_last_unsorted_element = index_at_which_to_split + (_nodes[node_index]).getNumTriangles() - 1;

    while (index_at_which_to_split <= index_of_last_unsorted_element) {
      Triangle<t> current_element = (*_triangulation)[_mesh_element_indices[index_at_which_to_split]];
      if ((current_element.getCentroid())[axis_index] < split_location) {
        index_at_which_to_split++;
      } else {
        this->swapElements(index_at_which_to_split, index_of_last_unsorted_element);
        index_of_last_unsorted_element--;
      }
    }

    return index_at_which_to_split; //! this is the index of the first element on the side > split_location
  }

  template <typename t>
  gpuify BVH<t>& BVH<t>::swapElements(unsigned int index_one, unsigned int index_two) {
    auto temporary = _mesh_element_indices[index_one];
    _mesh_element_indices[index_one] = _mesh_element_indices[index_two];
    _mesh_element_indices[index_two] = temporary;
    return *this;
  }

  template <typename t>
  gpuify unsigned int BVH<t>::createChildNodes(unsigned int node_index, unsigned int split_index, unsigned int num_triangles_on_left) {
    unsigned int left_child_index = this->getNumNodesUsed();

    this->setNumNodesUsed(left_child_index + 2);

    (_nodes[node_index]).setChildOneIndex(left_child_index);

    (_nodes[left_child_index]).setFirstTriangleIndex((_nodes[node_index]).getFirstTriangleIndex());
    (_nodes[left_child_index]).setNumTriangles(num_triangles_on_left);

    (_nodes[left_child_index + 1]).setFirstTriangleIndex(split_index);
    (_nodes[left_child_index + 1]).setNumTriangles((_nodes[node_index]).getNumTriangles() - num_triangles_on_left);

    (_nodes[node_index]).setNumTriangles(0);

    this->constructNewNode(left_child_index);
    this->constructNewNode(left_child_index + 1);

    return left_child_index;
  }

  template <typename t>
  gpuify BVH<t>& BVH<t>::constructNewNode(unsigned int node_index) {
    auto submesh = _triangulation->getSubMesh(this->getSubMeshIndices(node_index));
    (_nodes[node_index]).growToIncludeTriangulation( submesh );
    return *this;
  }

  template <typename t>
  gpuify std::vector<unsigned int> BVH<t>::getSubMeshIndices(unsigned int node_index) const {
    std::vector<unsigned int> element_array_indices = (_nodes[node_index]).getElementArraySubindices();
    std::vector<unsigned int> submesh_indices(element_array_indices.size());
    for (int i = 0; i < element_array_indices.size(); i++) {
      submesh_indices[i] = _mesh_element_indices[element_array_indices[i]];
    }
    return submesh_indices;
  }

  template <typename t>
  gpuify BVH<t>& BVH<t>::intersectRayWithBVHNode(std::shared_ptr<Ray<t>> ray, unsigned int node_index) {
    bool ray_intersects_with_node_bounding_box = (_nodes[node_index]).intersectRayWithNodeBoundingBox(ray);
    if (!ray_intersects_with_node_bounding_box) { return *this; }
    if ((_nodes[node_index]).isLeaf()) {
      for (unsigned int i = 0; i < (_nodes[node_index]).getNumTriangles(); i++) {
        ray->triangleIntersection((*_triangulation)[_mesh_element_indices[i + (_nodes[node_index]).getFirstTriangleIndex()]]);
      }
    } else {
      this->intersectRayWithBVHNode(ray, (_nodes[node_index]).getChildOneIndex());
      this->intersectRayWithBVHNode(ray, (_nodes[node_index]).getChildTwoIndex());
    }
    return *this;
  }

template class BVH<float>;
template class BVH<double>;
}